docs/proposals/maintenance.md

```diff
@@ -0,0 +1,179 @@
+face of crashes or restarts, but we do meet the needs of admins who want
+something more sophisticated than `sleep`.
+
+### Implementation of the above as a controller
+
+The parameters of a specific lame, drain, or unlame request could be pushed to
+the API server as a drain object.  The drain controller would have to manage
+multiple drains in flight at once.  If the drains have differing parameters,
+then more aggressive drains would generally starve safer ones.  While this could
+be a useful feature in emergencies, it will be a lot easier to reason about
+multiple drains in progress if they all share the same parameters, which would
+be configured via kube-env and would represent the cluster admin's SLO.
+
+The drain object would thus look something like this
+
+```
+type DrainSpec struct {
+  LameNodes    PodList
+  RemoveNodes  PodList
+  AddNodes     PodList
+  GracePeriod  int
+}
+```
+
+A drain object would work much like a deployment.  A drain controller would
+watch for drain objects and, within constraints, work toward the desired end
+state.  Nodes in `LameNodes` would merely be lamed, nodes is `RemoveNodes` would
+be drained according to the given parameters, and nodes in `AddNodes` would be
+undrained.  The object would be rejected if the sets overlap at all.  The action
+of the drain controller can be paused or stopped by deleting the drain object,
+and restarted by re-posting it.
+
+Undo would be trickier, but a clever client could be written such that it
+records the starting state of all the nodes and is able to send a reversing
+drain request.
+
+A central drain controller would have several advantages: safe restarts and the
+ability to react to other events in order to preserve the desired SLO.  For
+example, if a machine not in the list of RemoveNodes crashes and takes out other
+pods in replication sets also impacted by the drain, it can hold off on
+evictions impacting that replication set until MinSecondsBetweenEvictions and
+MinShardStrength are OK again.
+
+## Implementation of safety logic
+
+The implementation is the same whether there is one drain in flight or several.
+The drain agent (either kubectl or the drain controller) must watch for
+replication controllers and pods, and it will watch all pods to notice any time
+there is a disruption to a ReplicaSet.  Because there is no reliable mechanism
+to replay past events, when the controller restarts, it must wait until it
+builds up enough history to establish that it can enforce the safety
+constraints.  The standard loop on draining machines will be:
+
+- Is it safe to perform any evictions now?  If so, do them.
+- Are any machines empty?  If so mark them done.
+- Are any drains done?  If so mark them done.
+- Calculate the soonest we can do another eviction.
+- Sleep until then, or until we receive a watch update.
+
+## Future
+
+### Varying SLOs
+
+We could implement multiple SLOs, either by allowing the cluster admin to specify
```

Comment:
What you describe in this paragraph is reasonable but I'd suggest mentioning it briefly also in the Service Quality section -- say that time between eviction and shard strength are cluster-wide properties not per-collection. You could also say that eviction notice is optional -- it defaults to the PodSpec.TerminationGracePeriodSeconds if it is not specified (either in the kubectl version or the object-on-the-server version).